libraries/AP_ExternalAHRS/AP_ExternalAHRS_MicroStrain7.cpp

```diff
@@ -0,0 +1,367 @@
+enum class FilterState {
+    GQ7_INIT = 0x01,
+    GQ7_VERT_GYRO = 0x02,
+    GQ7_AHRS = 0x03,
+    GQ7_FULL_NAV = 0x04
+};
+
+AP_ExternalAHRS_MicroStrain7::AP_ExternalAHRS_MicroStrain7(AP_ExternalAHRS *_frontend,
+        AP_ExternalAHRS::state_t &_state): AP_ExternalAHRS_backend(_frontend, _state)
+{
+    auto &sm = AP::serialmanager();
+    uart = sm.find_serial(AP_SerialManager::SerialProtocol_AHRS, 0);
+
+    baudrate = sm.find_baudrate(AP_SerialManager::SerialProtocol_AHRS, 0);
+    port_num = sm.find_portnum(AP_SerialManager::SerialProtocol_AHRS, 0);
+
+    if (!uart) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "MicroStrain7 ExternalAHRS no UART");
+        return;
+    }
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_ExternalAHRS_MicroStrain7::update_thread, void), "AHRS", 2048, AP_HAL::Scheduler::PRIORITY_SPI, 0)) {
+        AP_BoardConfig::allocation_error("MicroStrain7 failed to allocate ExternalAHRS update thread");
+    }
+
+    hal.scheduler->delay(5000);
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "MicroStrain7 ExternalAHRS initialised");
+}
+
+void AP_ExternalAHRS_MicroStrain7::update_thread(void)
+{
+    if (!port_open) {
+        port_open = true;
+        uart->begin(baudrate);
+    }
+
+    while (true) {
+        build_packet();
+        hal.scheduler->delay_microseconds(100);
+    }
+}
+
+
+
+// Builds packets by looking at each individual byte, once a full packet has been read in it checks the checksum then handles the packet.
+void AP_ExternalAHRS_MicroStrain7::build_packet()
+{
+    if (uart == nullptr) {
+        return;
+    }
+
+    WITH_SEMAPHORE(sem);
+    uint32_t nbytes = MIN(uart->available(), 2048u);
+    while (nbytes--> 0) {
+        uint8_t b;
+        if (!uart->read(b)) {
+            break;
+        }
+        DescriptorSet descriptor;
+        if (handle_byte(b, descriptor)) {
+            switch (descriptor) {
+            case DescriptorSet::IMUData:
+                post_imu();
+                break;
+            case DescriptorSet::GNSSData:
+            case DescriptorSet::GNSSRecv1:
+            case DescriptorSet::GNSSRecv2:
+                break;
+            case DescriptorSet::FilterData:
+                post_filter();
+                break;
+            case DescriptorSet::BaseCommand:
+            case DescriptorSet::DMCommand:
+            case DescriptorSet::SystemCommand:
+                break;
+            }
+        }
+    }
+}
+
+
+
+// Posts data from an imu packet to `state` and `handle_external` methods
+void AP_ExternalAHRS_MicroStrain7::post_imu() const
+{
+    {
+        WITH_SEMAPHORE(state.sem);
+        state.accel = imu_data.accel;
+        state.gyro = imu_data.gyro;
+
+        state.quat = imu_data.quat;
+        state.have_quaternion = true;
+    }
+
+    {
+        AP_ExternalAHRS::ins_data_message_t ins {
+accel: imu_data.accel,
```

Comment:
Any idea why astyle is ruining the indent here? This is what astyle is automatically doing with the code.